File: day76/addBinary.py
"""
67. 二进制求和
给你两个二进制字符串，返回它们的和（用二进制表示）。
输入为 非空 字符串且只包含数字 1 和 0。

链接: https://leetcode-cn.com/problems/add-binary/
"""

# 二进制转十进制，相加后再转回二进制
# class Solution:
#     def addBinary(self, a, b) -> str:
#         return '{0:b}'.format(int(a, 2) + int(b, 2))


# 逐位相加
# class Solution:
#     def addBinary(self, a: str, b: str) -> str:
#         n = max(len(a), len(b))
#         a = a.zfill(n); b = b.zfill(n)
#         carry = 0
#         res = []
#         for i in range(n - 1, -1, -1):
#             cur = int(a[i]) + int(b[i]) + carry
#             carry = cur >> 1
#             res.append(str(cur % 2))
#         if carry == 1:
#             res.append("1")
#         return "".join(res[::-1])


# 位运算
class Solution:
    def addBinary(self, a, b) -> str:
        x, y = int(a, 2), int(b, 2)
        while y:
            x, y = x ^ y, (x & y) << 1
        return bin(x)[2:]

File: day76/addBinary.java
class Solution {
    public String addBinary(String a, String b) {
        int n = a.length(), m = b.length();
        if  (n > m) return addBinary(b, a);

        StringBuilder res = new StringBuilder();
        int carry = 0;
        for (int i = m - 1; i >= 0; i--) {
            int c = n > 0 && a.charAt(--n) == '1' ? 1:0;
            int d = b.charAt(i) == '1' ? 1:0;
            int cur = c + d + carry;
            carry = cur / 2;
            res.append(cur % 2 == 1? '1':'0');
        }
        if (carry == 1) res.append('1');
        res.reverse();
        return res.toString();
        
    }
}

// class Solution {
//   public String addBinary(String a, String b) {
//     return Integer.toBinaryString(Integer.parseInt(a, 2) + Integer.parseInt(b, 2));
//   }
// }
